### PR TITLE
Allow for passing country model in place of ID

### PR DESCRIPTION
**Pull request by dshoreman**
In most cases, where someone's calling `State::getNameList()`, they probably already have a `hasMany` countries relation setup on their model. With this commit, they will be able to pass the model into the method directly, instead of having to pass `$this->country->id`

Example:
```php
<?php namespace Acme\Foo\Models

use RainLab\Location\Models\State;

class Thing extends Model
{
    public $belongsTo = [
        'country' => ['RainLab\Location\Models\Country'],
        'state'   => ['RainLab\Location\Models\State'],
    ];

    public function getStateOptions()
    {
        // Passing the ID (the old way, still works)
        return State::getNameList(42);

        // Passing the ID of related Country (works sometimes, see below)
        return State::getNameList($this->country->id);

        // Passing the model (new way, easier, identical result)
        return State::getNameList($this->country);
    }
}
```

Currently the pay plugin attempts to pass `$this->country_id` but that's actually `NULL`.

With fields generally named by the respective relation (e.g. using `country` as opposed to `country_id`), we already have an instance of the Country model. It makes sense to take advantage of that fact, which is exactly what this PR does.

### Caveats when using `$this->country->id`
When first viewing Create page, it throws a `Trying to get property of non-object` error. As soon as a country is selected from the dropdown it fixes itself.

Passing model with `$this->country` works all the time, even prior to setting the country. With that in mind, best option for users would be a hard-coded integer or the related country model.

### CODE REVIEW

**Comment by dshoreman** [comment]
Update: Might need to hold off on this one for a bit, some strange things are happening in my testing.

In the OP, I mentioned `$this->country_id` is `NULL`, but there appears to be more to it.

Using the pay plugin as an example, if I create an invoice and go straight to Billing Details to set the Country, everything works ok. However, if I select a User first, all methods of passing the country evaluate to `NULL` resulting in the State list not getting populated. I tried removing `dependsOn: user` from the country field, but the same thing happens.

In short... It would seem that selecting a User overrides the country ID that is set in the form, even when the User doesn't have a country set. This in turn breaks the State dropdown as there is no country value for it to search by.

**Comment by daftspunk** [comment]
Should I merge this?

**Comment by dshoreman** [comment]
Completely forgot about this issue, thanks for the reminder.

The short answer is *not yet*. This PR may need more work, I need to double check the issue I had previously. I'll be working on addresses again in Octoshop shortly, hoping to finish that part fairly soon.

Since this PR affects the work on addresses, I'll check it then and ping you when we're good to go.

**Comment by daftspunk** [comment]
Looks like I can't merge it due to #2 anyways... zzzz
